src/database/mod.rs

```diff
@@ -130,229 +126,103 @@ impl Database {
+                       VALUES ($1, (SELECT COALESCE(MAX(leaf_index) + 1, 0) FROM identities), $2)
+                       ON CONFLICT DO NOTHING
+                       RETURNING leaf_index;"#,
         )
-        .bind(group_id as i64)
         .bind(identity)
-        .bind(IDENTITY_PENDING);
-        self.pool.execute(query).await?;
-        Ok(())
+        .bind(<&str>::from(Status::Pending));
+        let row = self.pool.fetch_optional(query).await?;
+        Ok(row.map(|row| row.get::<i64, _>(0) as usize))
     }
 
-    pub async fn start_identity_insertion(
+    pub async fn mark_identity_submitted_to_contract(
         &self,
-        group_id: usize,
-        commitment: &Hash,
+        leaf_index: usize,
     ) -> Result<(), Error> {
         let query = sqlx::query(
-            r#"UPDATE pending_identities
-                   SET status = $4
-                   WHERE group_id = $1 AND commitment = $2 AND status = $3;"#,
+            r#"UPDATE identities
+                       SET status = $2
+                       WHERE leaf_index = $1;"#,
         )
-        .bind(group_id as i64)
-        .bind(commitment)
-        .bind(IDENTITY_PENDING) // previous status
-        .bind(IDENTITY_SUBMITTING); // new status
-
+        .bind(leaf_index as i64)
+        .bind(<&str>::from(Status::Mined));
         self.pool.execute(query).await?;
         Ok(())
     }
 
-    pub async fn mark_identity_inserted(
+    pub async fn get_updates_range(
         &self,
-        group_id: usize,
-        commitment: &Hash,
-    ) -> Result<(), Error> {
+        from_index: usize,
+        to_index: usize,
+    ) -> Result<Vec<TreeUpdate>, Error> {
         let query = sqlx::query(
-            r#"UPDATE pending_identities
-                   SET status = $3
-                   WHERE group_id = $1 AND commitment = $2;"#,
+            r#"SELECT commitment, leaf_index, status
+                       FROM identities
+                       WHERE leaf_index >= $1 AND leaf_index <= $2
+                       ORDER BY leaf_index ASC;"#,
         )
-        .bind(group_id as i64)
-        .bind(commitment)
-        .bind(IDENTITY_MINED);
-
-        self.pool.execute(query).await?;
-        Ok(())
+        .bind(from_index as i64)
+        .bind(to_index as i64);
+        let rows = self.pool.fetch_all(query).await?;
+        rows.iter()
+            .map(|row| {
+                let element = row.get::<Hash, _>(0);
+                let leaf_index = row.get::<i64, _>(1) as usize;
+                Ok(TreeUpdate {
+                    leaf_index,
+                    element,
+                })
+            })
+            .collect::<Result<Vec<_>, _>>()
     }
 
-    pub async fn delete_pending_identity(
-        &self,
-        group_id: usize,
-        commitment: &Hash,
-    ) -> Result<(), Error> {
+    pub async fn get_identity_index(&self, identity: &Hash) -> Result<Option<TreeItem>, Error> {
```

Comment:
This name is also a bit strange. Am I right in understanding that it gets the index in the tree of the provided `identity`? If that's the case, perhaps `get_identity_index_in_tree`?

Reply:
changed to `get_identity_leaf_index`, leaf index is what we consistently use for this